src/main/scala/org/constellation/domain/redownload/RedownloadService.scala

```diff
@@ -420,20 +456,129 @@ class RedownloadService[F[_]: NonEmptyParallel: Applicative](
+            )
+          ),
+        if (currentRequests.values.toList.distinct.size == 1 && currentRequests.size == 3)
+          redownloadStorage.clearJoinActivePoolCommands() >>
+            //joinActivePoolCommandRequests.modify(_ => (Map.empty, ())) >>
+            logger.debug(s"Added joinActivePool request and the condition to join active pool was met!")
+        else
+          F.raiseError[Unit](
+            new Throwable(
+              s"Added joinActivePool request but the amount of needed requests wasn't met. request=$joinActivePoolCommand! currentRequests size = ${currentRequests.size} != 3."
+            )
+          )
+      )
+    } yield ()
+
+  def redownloadBeforeJoiningActivePeersPool(senderId: Id, joinActivePoolCommand: JoinActivePoolCommand): F[Unit] = {
+    for {
+      _ <- addAndCheckJoinActivePoolCommand(senderId, joinActivePoolCommand)
+      JoinActivePoolCommand(lastActiveFullNodes, lastActiveBetweenHeight) = joinActivePoolCommand
+      _ <- logger.debug(s"Joining active peers pool! lastActivePeers: ${logIds(lastActiveFullNodes)}")
+      _ <- logger.debug(s"Joining active peers pool! lastActiveBetweenHeight: $lastActiveBetweenHeight")
+      peers <- lastActiveFullNodes.toList
+        .traverse(
+          clusterStorage.getPeer
+        )
+        .map(_.flatten.map(_.copy(majorityHeight = NonEmptyList.one(lastActiveBetweenHeight)))) // TODO: should we handle the situation when at least one of the last active peers is not found?
+      _ <- logger.debug(s"Joining active peers pool! peers=$peers")
+      _ <- clear
+      _ <- fetchAndUpdatePeersProposals(peers.some)
+      _ <- checkForAlignmentWithMajoritySnapshot(isJoiningActivePool = true, lastActivePeers = peers.some)
+      latestMajorityHeight <- redownloadStorage.getLatestMajorityHeight
+      _ <- logger.debug(s"Joining active peers pool! latestMajorityHeight: $latestMajorityHeight")
+      // TODO: this was in snapshotService
+      newActivePeers <- snapshotServiceStorage.getNextSnapshotFacilitators
+      _ <- logger.debug(s"Joining active peers pool! newActivePeers: $newActivePeers") //TODO: logging function
+      activeBetweenHeights = Cluster.calculateActiveBetweenHeights(latestMajorityHeight, activePeersRotationEveryNHeights)
+      _ <- clusterStorage.setActivePeers(newActivePeers, activeBetweenHeights, metrics)
+    } yield ()
+  }.handleErrorWith { e =>
+    logger.error(e)("Error during joining active peers pool!")
+  }
+
+  // TODO: notify new light nodes to join the pool also and update your light nodes list
+  // TODO: we need a consensus over new light and full nodes between current active full nodes
+  private def notifyNewActivePeersAndLeaveThePool(): F[Unit] =
+    for {
+      _ <- logger.debug(s"Notifying new active peers to join the active pool!")
```

Comment:
Why this is a part of `redownload service`? It should be separated

Reply:
I guess it can be moved. Redownload flow seemed like the best place to put it initially.

Reply:
imo it should be in a separate service. and the same for storage related to joining pool